### ISSUE TITLE
Sealevel: Kathy messages are sent to/from sealevel chains

### ISSUE THREAD

**Issue by tkporter** [hyperlane-xyz/hyperlane-monorepo#2503]
Partially blocked by https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/2502 - I think we can start this ticket before that's finished though

* We decided in https://discord.com/channels/935678348330434570/935678739663192184/1126547994884460634 to keep this TS native
* We need to decide if we want to do some SDK-level refactoring to accommodate this, or if we should just keep the SDK mostly as-is and shove the sealevel stuff in the Kathy script
* Note key-wise, I don't 100% recall if kathy uses a hexKey or AWS KMS key, but sealevel requires hexKey. Take a look at what was done for the relayer to accommodate using AWS KMS keys for Ethereum & hexKey for sealevel
* Expectation is that Kathy can be ran in testnet3 (or mainnet2 too, if we have Sealevel deployments by the time work is being done here), and messages will be sent to & from sealevel chains

```[tasklist]
### Tasks
- [ ] https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/2587
- [ ] https://github.com/hyperlane-xyz/issues/issues/558
```

**Comment by jmrossy** [comment]
# Sealevel support proposal

## Overview

We've so far taken a somewhat top-down approach to adding Sealevel support; the warp UI, hyp-deploy have been updated before the SDK and Utils packages. We can continue that trend with HelloWorld/Kathy or we can make a bigger upfront investment and redesign the SDK+Utils now.

## Option 1: Shortest path

Kathy is built around HelloWorld, which is just a thin wrapper around the App/Deployer/Checker abstractions in the SDK. So updating _just_ HelloWorld to be sealevel compatible wouldn't make sense. Sealevel support for Kathy would require a new, parallel HelloWorld version that live alongside the EVM one.

This option is still less work than Option 2 but feels short-sighted.

## Option 2: Multi-env SDK

Updating the common utilities (e.g. address encoder, msg formatter) and SDK primitives (e.g. MultiProvider, App, Checker, Deployer) feels like a daunting task but also seems like the natural next step for Sealevel support.

I won't lay out a complete plan just yet until we have agreement on this direction but structural changes would include:

- Create multi-env compatible address/msg utilities (some of this is done in warp UI already)
- Redesign MultiProvider to handle different provider/connection libs
- Decouple all SDK primitives from Ethers v5 contracts/factories/providers
  - Separate implementation + interface for current classes
  - Build Sealevel equivalent impls
- Update/extend tests to cover Sealevel too (currently hardhat-based)

The new architecture would ideally be a plugin-like architecture with sub-packages. Something that allows SDK users to bundle only the libs for the environments they care about. Otherwise all EVM-only users would be unintentionally including in large libs for Sealevel (and soon Cosmos).

IMHO, this option is tricky but feasible.

**Comment by jmrossy** [comment]
Continuing on the last post above ^, we decided on Discord to try option 1.5. We will not modify existing classes in SDK but we will add new multi-protocol versions (e.g. `MultiProtocolProvider`) as needed to get a HelloWorld (and then Kathy) working.

So far I've got a [Draft PR](https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/2529/files#diff-eaeb30ae5ebadd7c49b753cbd99c41fb72e67eb939c444fbcdc1b635e1a962de) open with a new MultiProtocolProvider based partly on the one I made for Warp UI. This introduces a new set of Provider-related types and builders, which allow us to support different libraries (Viem, Ethers6, SolanaWeb3, etc.)

Next I took a stab at a `MultiProtocolApp` abstraction today but sadly the concept falls apart almost immediately 😭 . Our `HyperlaneApp` is a fancy map, from chain -> contracts. Sealevel's JS lib (`solana/web3.js`) doesn't have anything comparable to an Ethers/Viem Contract. Data must be manually (de)serialized with serializers custom defined for each contract. So scrap `MultiProtocolApp` I think.

Children Apps like RouterApp add some functionality like fetching ISMs or quoting gas. We'll def still want those features somewhere. Another step down, TokenApp adds a transfer method, i.e. a method to translate a user intent into a tx. Like @yorhodes  pointed out last week, this is where the TokenApp overlaps with my ITokenAdapter in the Warp UI:

`TokenApp = ITokenAdapter + Ethers Contracts`

I'm not convinced that the encapsulation of state (in this can contracts + addresses) with the logic to use those contracts simplifies things here. And we don't have contracts for Sealevel anyway 🤷‍♂️ . I guess that leaves us with just adapters for translating intent into txs across different provider types. Not sure yet how this should look in the SDK. I'll continue tmrw
